<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
</head>
<body>
    <script>
        
        let usuarios = ['Adriano', 'Marcia', 'José']

        function inserirUsuario(nome){

            let promise = new Promise(function(resolve, reject){//tem esses 2 argumentos.

                setTimeout(()=>{ //usado para simular o delay
                usuarios.push(nome)
               
                    let error = false //Para se der erro> Se quiser testar com erro, é só colocar true aqui

                    if (!error){ //esse ! significa que não teve erro
                        resolve()
                    }else{
                        reject({msg:'Erro de qualquer coisa'})
                    }
                }, 1000)
            })
            return promise
   
        }

        function listarUsuarios(){
            console.log(usuarios)
        }

        async function executar(){ //await só funciona dentro de função async
            await inserirUsuario('Igor') /* await vai fazer esperar inserir o usuário (promise ser resolvida) para poder listar os usuários. Caso não tivesse o await, ele ia imprimir a lista sem igor, pq embora o comando primeiro mande incluir igor, a função inserir lá em cima demora 1seg para inserir, então ele não espera teminar e vai logo listando o que já tinha antes */
            listarUsuarios()
        }
        executar()

    </script>
</body>
</html>